SCons/Tool/ninja.py

```diff
@@ -0,0 +1,1570 @@
+        )
+
+    # pylint: disable=too-many-branches
+    def handle_list_action(self, node, action):
+        """TODO write this comment"""
+        results = [
+            self.action_to_ninja_build(node, action=act)
+            for act in action.list
+            if act is not None
+        ]
+        results = [
+            result for result in results if result is not None and result["outputs"]
+        ]
+        if not results:
+            return None
+
+        # No need to process the results if we only got a single result
+        if len(results) == 1:
+            return results[0]
+
+        all_outputs = list({output for build in results for output in build["outputs"]})
+        dependencies = list({dep for build in results for dep in build["implicit"]})
+
+        if results[0]["rule"] == "CMD":
+            cmdline = ""
+            for cmd in results:
+
+                # Occasionally a command line will expand to a
+                # whitespace only string (i.e. '  '). Which is not a
+                # valid command but does not trigger the empty command
+                # condition if not cmdstr. So here we strip preceding
+                # and proceeding whitespace to make strings like the
+                # above become empty strings and so will be skipped.
+                cmdstr = cmd["variables"]["cmd"].strip()
+                if not cmdstr:
+                    continue
+
+                # Skip duplicate commands
+                if cmdstr in cmdline:
+                    continue
+
+                if cmdline:
+                    cmdline += " && "
+
+                cmdline += cmdstr
+
+            # Remove all preceding and proceeding whitespace
+            cmdline = cmdline.strip()
+
+            # Make sure we didn't generate an empty cmdline
+            if cmdline:
+                ninja_build = {
+                    "outputs": all_outputs,
+                    "rule": "CMD",
+                    "variables": {
+                        "cmd": cmdline,
+                        "env": get_command_env(node.env if node.env else self.env),
+                    },
+                    "implicit": dependencies,
+                }
+
+                if node.env and node.env.get("NINJA_POOL", None) is not None:
+                    ninja_build["pool"] = node.env["pool"]
+
+                return ninja_build
+
+        elif results[0]["rule"] == "phony":
+            return {
+                "outputs": all_outputs,
+                "rule": "phony",
+                "implicit": dependencies,
+            }
+
+        elif results[0]["rule"] == "INSTALL":
+            return {
+                "outputs": all_outputs,
+                "rule": "INSTALL",
+                "inputs": [get_path(src_file(s)) for s in node.sources],
+                "implicit": dependencies,
+            }
+
+        raise Exception("Unhandled list action with rule: " + results[0]["rule"])
```

Comment:
Why does the handle_list_action function only check the rule of the first result for its list of results? It calls the action_to_ninja_build function for each item which may return a different ninja rule for each item in results. 

I am understanding this function to 

1. take in a list action and convert each action in the list to a ninja build dictionary
2. don't actually use the series of dictionary in those results, but generate a new single ninja build dictionary to represent all of them

So if we only check the first item we, are making some assumptions that the rest are of that same type?

Could instead the several ninja build dictionaries be made to depend on each other so they execute in the same order? Maybe some of the current handling are optimization cases where its better to have a single build item, but those could still be handled separately, and instead of just raising exception at the end here we generate all the ninja build items and make them depend on each other?

Reply:
> Could instead the several ninja build dictionaries be made to depend on each other so they execute in the same order? 

This is how it used to work and it was too hard to get a ninja file that both functioned properly (had correct DAG) and that Ninja would happily parse. In the interest of shipping I changed this to the simpler solution here which doesn't break on the Mongo code base (which was our smaller subset of caring at the time).

tl;dr It can be done that way but it was more fiddly than I had time for when originally writing this tool.